### ISSUE TITLE
Calculate Log Probabilities for a Sequence of Speech Tokens Given Text Tokens

### ISSUE THREAD

**Issue by ryanxai** [WhisperSpeech/WhisperSpeech#153]
While I'm really thankful that you've open-sourced this amazing repo, I'm a bit confused about the way you are generating speech tokens from text tokens. It seems that the generate function which is defined in t2s_up_wds_mlang_enclm.py is feeding only one speech token to the forward method to generate the next speech token while the generate_batch which is defined in the same file (and it's not working) it uses all speech frames up to a specific time to generate the next token. Based on what I described I have two questions:

1- Why is generate function only using one speech token as input for the forward method?
2- How can I make the forward method work with all speech and text tokens fed to it? (currently it doesn't work this way and only accepts one speech token)

The reason I need item 2 is that I want to calculate the log probabilities for all speech tokens given text tokens.

Thanks in advance.

**Comment by BBC-Esq** [comment]
Great question! This is actually about **KV-caching optimization**, not a limitation of the model architecture.

## What's happening

Looking at the `generate` method, there are actually two phases:

```python
# Phase 1: "Prefill" - processes ALL tokens up to start+1
with record_function("prefill"):
    toks[:,start+1] = self.generate_one(toks[:,:start+1].contiguous(), toks_positions[:start+1], ...)

# Phase 2: Incremental generation - only passes the LAST token
with inference.inference_context():
    for i in it:
        toks[:,i+1] = self.generate_next(toks[:,i:i+1], toks_positions[i:i+1], ...)
```

The reason only one token is passed in the loop is that `optimize()` sets up **KV-caching** in the attention layers (see `setup_kv_cache` in `modules.py`). The key/value projections for previous tokens are stored in the cache, so the model only needs to compute attention for the new token while still attending to all previous positions.

## Computing log probabilities for all tokens

To get logits for a full sequence (teacher-forcing style), you can use the `forward` method directly **without** calling `optimize()`:

```python
import torch
import torch.nn.functional as F
from whisperspeech2.t2s_up_wds_mlang_enclm import TSARTransformer
from whisperspeech2 import languages

# Load model WITHOUT optimize()
model = TSARTransformer.load_model(device="cuda")
# model.optimize()  # Don't call this!

# Prepare inputs
text = "Hello world"
lang = "en"
cps = 15

# Tokenize text
model.ensure_tokenizer()
ttoks = torch.tensor(model.tokenizer.encode(text), device=model.device)
ttoks = F.pad(ttoks, (1, model.ttoks_len - len(ttoks) - 1), value=model.tokenizer.eot).unsqueeze(0)

langs = torch.tensor([languages.to_id(lang)], device=model.device)
cpss = torch.tensor([cps], device=model.device)

# Assume you have ground truth speech tokens `stoks` of shape (seq_len,)
# Prepend the BOS/padding token
bos = model.stoks_codes + model.tunables.padding_token_offset
in_stoks = F.pad(stoks, (1, 0), value=bos).unsqueeze(0)  # (1, seq_len+1)
out_stoks = F.pad(stoks, (0, 1), value=bos).unsqueeze(0)  # targets shifted

# Create positions for the full sequence
in_stoks_positions = torch.arange(in_stoks.shape[1], device=model.device)

# Forward pass - get logits for ALL positions
with torch.no_grad():
    logits, loss = model(
        in_ttoks=ttoks,
        out_ttoks=ttoks,  # for encoder loss if causal_encoder=True
        languages=langs,
        cpss=cpss,
        in_stoks=in_stoks,
        out_stoks=out_stoks,
        in_stoks_positions=in_stoks_positions,
        loss=True
    )

# logits shape: (batch, seq_len, vocab_size)
# Compute log probabilities
log_probs = F.log_softmax(logits, dim=-1)

# Get log prob for each ground truth token
token_log_probs = log_probs.gather(-1, out_stoks.unsqueeze(-1)).squeeze(-1)
```

The key points:
1. Don't call `optimize()` since that enables KV-caching which expects incremental generation
2. Pass full `in_stoks_positions` as `torch.arange(0, seq_len)`
3. The model will return logits for every position

Let me know if you run into any issues!
